### PR TITLE
Requests cleaned up and refreshed

### LINKED ISSUE #402: Pull member elements on selection-based pull of Groups
#### Description:
When using a `SelectionFilterRequest` and pulling a Revit group, only a `ModelInstance` correspondent to this group is pulled. What is more, when the `FilterRequest` is combined with type filter request, nothing gets pulled.

#### Test files:
Issue-specific test .gh file has been uploaded to [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue402%2DPullMemberElementsOfGroupsAndAssemblies). Example of groups can be found in Beams.rvt file in [this folder](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPull%2FPhysical).

### LINKED ISSUE #610: Revit_Toolkit: Add useful requests
#### Description:

A few more requests could be added:
- `FilterElementsOnly `
- `FilterTypesOnly `
- `FilterFamilyTypeByName` - to be investigated together with `ElementIdsByFamilyAndType` - possibly the if/else at the end of the latter could be eliminated in that go

### LINKED ISSUE #684: Revit Pull Methods for Facilitating LCA
#### Fuctionality:
In order to make our LCA-toolkit easily usable for projects that have Revit models, I suggest we set up a way to automatically sift through the endless Revit families (circled in red), identify which are in use in the Revit file, and pull and organize these geometries by unique family types (lets pretend those in use are highlighted in yellow). The family type is the level of detail I'd imagine working at in Revit. Perhaps this belongs in Revit-Toolkit but i don't know for sure so I kept it here for now!

image

A rough sketch of such a concept is below, where when set to true the left side of the component populates with all the families in use in an open Revit model. Then the user can set each of those families to TRUE in order to confirm which families they'd like to pull the geometry from. The right side of the component would then spit out the geometry from all families which are set to TRUE and auto- organizes these into the unique family types. I realise this isn't a realistic component design given how the RevitAdapter operated but I hope this illustrates my point. 

Revit Pull sketch

Given how the RevitAdapter is set up I imagine what I'm really looking for is actually two methods:
1. A method that returns the active Families in Revit which have objects in them, allows the user to select/confirm which they wish to pull from a list and constructs an appropriate FilterRequest for use in the inputs for the RevitPull. 
2. A method that takes the output from the RevitPull geometry, groups and organizes it into family types for easy use specifying.

### LINKED ISSUE #695: Revit_Toolkit: implement FilterFamilyTypesByUsage
#### Description:

As in the title, it would be nice to be able to pull all family types that are used/unused in the model.

### LINKED ISSUE #702: Revit_Toolkit: Introduce MemberElementsRequest
#### Description:

To allow querying e.g. Groups, Assemblies or MEPSystems for their member elements.

### LINKED ISSUE #705: Revit_Toolkit: FilterByFamilyAndTypeName works only for chosen types
#### Description:

This is due to the fact that `if (type == null) continue;`

https://github.com/BHoM/Revit_Toolkit/blob/cf80a0e17e72720a9cfce2a782265b794dc75922/Engine_Revit_UI/Query/ElementIds/ElementIdsByFamilyAndType.cs#L90-L131

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Closes #402
Closes #610
Closes #684
Closes #695
Closes #702
Closes #705

### Test files

- [Macro test file](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPull%2FRevit%2FRequests)
- [#402](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue402%2DPullGroupMembers)
- [#610](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue610%2DUsefulRequests)
- [#684](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue684%2DPullForLCA)
- [#695](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue695%2DFilterByUsage)
- [#702](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue702%2DFilterMemberElements) - this is likely to need more tests, would be nice to prepare sample test files for each MEP subdiscipline @kayleighhoude 
- [#705](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue705%2DFilterByFamilyAndTypeNameBug)
- [Pulling element types](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DPullType) - good started to updating type properties @kayleighhoude 

### Changelog

- Selection-based pull has been fixed to include member elements of groups (and groups in groups)
- `FilterModelElements` request has been added
- `FilterFamilyTypeByName` request has been added
- Filtering types has been enabled by using `FilterFamilyTypeByName` with blank string inputs
- `FilterByUsage` request has been added
- `FilterByFamilyType` request has been added
- `FilterMemberElements` request has been added
- a `CustomData` key `Revit_familyTypeId` has been added for easier query of element types
- `FilterFamilyTypesOfFamily` has been renamed to `FilterTypesOfFamily`
- `IsAnalytical`, `IsUsed` and `InstanceType` query methods have been added to support the above
- a few minor bugs were fixed and a few method slightly refactored

### Additional comments

Sorry for a monster PR, but it was more convenient to tackle multiple issues in one go as they often were interconnected - the test files are as granular as possible to make the review process digestible.

### CODE REVIEW

**Comment by pawelbaran** [comment]
Thanks @kThorsager - I put a do-not-merge label until the Grid/Level pull is fixed.

> Accidentally tried pulling from Revit2019 with Debug2020 built, and got this error:
1. Failed to pull. Exception from the adapter: Could not load type 'BH.oM.Adapters.Revit.Requests.FilterFamilyTypesOfFamily' from assembly 'Revit_oM, Version=3.0.0.0, Culture=neutral, PublicKeyToken=null'.
Which isn't super informative, not necessarily related to this PR, but maybe worth mentioning.

It is related to this PR, but the error was caused by the fact that there is a single Revit_oM assembly serving 3 version-specific Engine_Revit_UI assemblies - you used Revit_oM from this branch with Engine_Revit_UI_2019 from master.

**Comment by kThorsager** [comment]
> Thanks @kThorsager - I put a do-not-merge label until the Grid/Level pull is fixed.

I have since been unable to reproduce this delay after testing forwards and backwards between master and this branch, so I would say that it was some strange local error unrelated to this PR.

File: Revit_Engine/Create/Requests/FilterMemberElements.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+{
+    public static partial class Create
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Creates an IRequest that filters elements being members of selection sets, assemblies, systems etc.")]
+        [Input("bHoMObject", "BHoMObject that contains ElementId of a correspondent Revit element under Revit_elementId CustomData key - usually previously pulled from Revit.")]
+        [Output("F", "IRequest to be used to filter views that implement a given template.")]
```
Comment (kThorsager) [inline Revit_Engine/Create/Requests/FilterMemberElements.cs:39]: Is this only for Views and Templates?

**Comment by pawelbaran** [comment]
Nice catch @kThorsager. It is fixed now, I also pushed myself to add missing descriptions in ElementId queries.

File: Engine_Revit_UI/Query/ElementIds/ElementIdsOfFamilyTypes.cs [hunk 2]
```diff
@@ -63,7 +63,14 @@ public static IEnumerable<ElementId> ElementIdsOfFamilyTypes(this Document docum
         }
 
         /***************************************************/
-        
+
+        [Description("Filters ElementIds of Revit family types based on names of theirs and their parent family, with option to loose the search by leaving one or both of the input names blank..")]
```
Comment (kThorsager) [inline Engine_Revit_UI/Query/ElementIds/ElementIdsOfFamilyTypes.cs:67]:
```suggestion
        [Description("Filters ElementIds of Revit family types based on names of theirs and their parent family, with option to loose the search by leaving one or both of the input names blank.")]
```

**Comment by pawelbaran** [comment]
Fix pushed @FraserGreenroyd, should be fine now :>

**Pull request merged**